src/goto-programs/string_refine_preprocess.cpp

```diff
@@ -0,0 +1,1390 @@
+      assignments.emplace_back(array_lhs, deref_data);
+      assignments.emplace_back(lhs, new_rhs);
+      insert_assignments(goto_program, i_it, assignments);
+      i_it=goto_program.insert_after(i_it);
+      pair.first->second=lhs;
+    }
+    return pair.first->second;
+  }
+  else if(rhs.id()==ID_typecast &&
+          is_java_string_pointer_type(rhs.op0().type()))
+  {
+    exprt new_rhs=make_cprover_string_assign(
+      goto_program, i_it, rhs.op0(), location);
+    return typecast_exprt(new_rhs, rhs.type());
+  }
+  else
+    return rhs;
+}
+
+/*******************************************************************\
+
+Function: string_refine_preprocesst::make_normal_assign
+
+  Inputs: a goto_program, a position in this program, an expression lhs,
+          a function type, a function name, a vector of arguments, a location
+          and a signature
+
+ Purpose: replace the current instruction by:
+          > lhs=function_name(arguments) : return_type @ location
+          If given, signature can force String conversion of given arguments.
+          The convention for signature is one character by argument
+          and 'S' denotes string.
+
+\*******************************************************************/
+
+void string_refine_preprocesst::make_normal_assign(
+  goto_programt &goto_program,
+  goto_programt::targett i_it,
+  const exprt &lhs,
+  const code_typet &function_type,
+  const irep_idt &function_name,
+  const exprt::operandst &arguments,
+  const source_locationt &location,
+  const std::string &signature)
+{
+  if(function_name==ID_cprover_string_copy_func)
+  {
+    assert(arguments.size()>0);
+    make_string_copy(goto_program, i_it, lhs, arguments[0], location);
+  }
+  else
+  {
+    function_application_exprt rhs(
+      symbol_exprt(function_name), function_type.return_type());
+    rhs.add_source_location()=location;
+    declare_function(function_name, function_type);
+
+    exprt::operandst processed_arguments=process_arguments(
+      goto_program, i_it, arguments, location, signature);
+    rhs.arguments()=processed_arguments;
+
+    code_assignt assignment(lhs, rhs);
+    assignment.add_source_location()=location;
+    i_it->make_assignment();
+    i_it->code=assignment;
+  }
+}
+
+/*******************************************************************\
+
+Function: string_refine_preprocesst::insert_assignments
+
+  Inputs: a goto_program, a position in this program, a list of assignments
+
+ Purpose: add the assignments to the program in the order they are given
+
+\*******************************************************************/
+
+void string_refine_preprocesst::insert_assignments(
+  goto_programt &goto_program,
+  goto_programt::targett &i_it,
+  const std::list<code_assignt> &va)
+{
+  if(va.empty())
+    return;
+
+  auto i=va.begin();
+  i_it->make_assignment();
+  i_it->code=*i;
+  for(i++; i!=va.end(); i++)
+  {
+    i_it=goto_program.insert_after(i_it);
```

Comment:
neither source_location nor function is properly set for these newly inserted instructions